crates/sui-rest-api/src/transactions/resolve.rs

```diff
@@ -0,0 +1,564 @@
+        );
+        resolved_transaction.gas_data_mut().budget = estimate;
+        estimate
+    };
+
+    // If we ended up using a mock gas coin we need to do gas selection now
+    if simulation_result.mock_gas_id.is_some() {
+        let input_objects = resolved_transaction
+            .input_objects()
+            .map_err(anyhow::Error::from)?
+            .iter()
+            .flat_map(|obj| match obj {
+                sui_types::transaction::InputObjectKind::ImmOrOwnedMoveObject((id, _, _)) => {
+                    Some(*id)
+                }
+                _ => None,
+            })
+            .collect_vec();
+        let gas_coins = select_gas(
+            &state.reader,
+            resolved_transaction.gas_data().owner,
+            budget,
+            protocol_config.max_gas_payment_objects(),
+            &input_objects,
+        )?;
+        resolved_transaction.gas_data_mut().payment = gas_coins;
+    }
+
+    // Do one last dry-run
+    let SimulateTransactionResult {
+        input_objects,
+        output_objects,
+        events,
+        effects,
+        ..
+    } = executor
+        .simulate_transaction(resolved_transaction.clone())
+        .map_err(anyhow::Error::from)?;
+
+    let events = events.map(Into::into);
+    let effects = effects.into();
+
+    let input_objects = input_objects
+        .into_values()
+        .map(Into::into)
+        .collect::<Vec<_>>();
+    let output_objects = output_objects
+        .into_values()
+        .map(Into::into)
+        .collect::<Vec<_>>();
+    let balance_changes = derive_balance_changes(&effects, &input_objects, &output_objects);
+
+    ResolveTransactionResponse {
+        transaction: resolved_transaction.into(),
+        events,
+        effects,
+        balance_changes: Some(balance_changes),
+        input_objects: Some(input_objects),
+        output_objects: Some(output_objects),
+    }
+    .pipe(|response| match accept {
+        AcceptFormat::Json => ResponseContent::Json(response),
+        AcceptFormat::Bcs => ResponseContent::Bcs(response),
+    })
+    .pipe(Ok)
+}
+
+struct NormalizedPackage {
+    #[allow(unused)]
+    package: MovePackage,
+    normalized_modules: BTreeMap<String, normalized::Module>,
+}
+
+fn called_packages(
+    reader: &StateReader,
+    protocol_config: &ProtocolConfig,
+    unresolved_transaction: &UnresolvedTransaction,
+) -> Result<HashMap<ObjectId, NormalizedPackage>> {
+    let binary_config = sui_types::execution_config_utils::to_binary_config(protocol_config);
+    let mut packages = HashMap::new();
+
+    for move_call in unresolved_transaction
+        .ptb
+        .commands
+        .iter()
+        .filter_map(|command| {
+            if let Command::MoveCall(move_call) = command {
+                Some(move_call)
+            } else {
+                None
+            }
+        })
+    {
+        let package = reader
+            .inner()
+            .get_object(&(move_call.package.into()))?
```

Comment:
Will this read objects from RocksDB serially, or is it doing something more clever behind the scenes? If the latter, is there a way we can perform all the fetches concurrently, and is that even worth doing?

Reply:
Yes reads are done sequentially I'm not overly concerned with perf on this part though, but we can easily optimize later